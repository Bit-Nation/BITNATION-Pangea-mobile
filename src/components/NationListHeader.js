import React, { Component } from 'react';
import {
  View,
  Text,
} from 'react-native';
import PropTypes from 'prop-types';
import { MediaQueryStyleSheet } from 'react-native-responsive';

export default class NationListHeader extends Component {

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.text}>
          {this.props.title}
        </Text>
      </View>
    );
  }

}

const styles = MediaQueryStyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    height: 30,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#EFEFF4',
    opacity: 0.5,
  },
  text: {
    flex: 1,
    marginLeft: 15,
    color: '#6D6D72',
    fontSize: 13,
  },
});
