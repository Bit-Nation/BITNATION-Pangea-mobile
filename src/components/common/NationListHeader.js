import React, { Component } from 'react';
import { View, Text, } from 'react-native';
import PropTypes from 'prop-types';
import { MediaQueryStyleSheet } from 'react-native-responsive';

import GlobalStyles from '../../global/Styles';

export default class NationListHeader extends Component {

  render() {
    return (
      <View style={styles.sectionListHeaderContainer}>
        <Text style={styles.sectionListHeaderText}>
          {this.props.title}
        </Text>
      </View>
    );
  }

}

NationListHeader.propTypes = {
  title: PropTypes.string,
};

const styles = MediaQueryStyleSheet.create({
  ...GlobalStyles,
});
