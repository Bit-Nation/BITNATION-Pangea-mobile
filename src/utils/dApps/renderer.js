// @flow

import React from 'react';

import Text from '../../components/dApps/Text';
import View from '../../components/dApps/View';
import TextInput from '../../components/dApps/TextInput';
import Button from '../../components/dApps/Button';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => {
  switch (typeName) {
    case 'Text':
      return Text;
    case 'View':
      return View;
    case 'TextInput':
      return TextInput;
    case 'Button':
      return Button;
    default:
      console.warn(`Unknown component type ${typeName}.`);
      return null;
  }
};

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, stateBased: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const stateBasedProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (propName.endsWith('Path') && validProps.stateBased.includes(propName.substring(0, propName.length - 4))) {
      // It's an allowed prop to set or get root component state. That properties are marked by Path suffix.
      stateBasedProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, stateBasedProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = getTypeElementFromText(type);

  if (component === null) {
    return null;
  }

  const { props } = json;

  const { nativeProps = props, stateBasedProps = {}, callbackProps = {} } =
    component.validNativeProps !== undefined
      ? validateProps(json.props, {
        native: component.validNativeProps,
        stateBased: Object.keys(component.stateBasedProps || {}),
        callbacks: component.callbackProps || [],
      }, type)
      : {};

  return React.createElement(
    component,
    { ...nativeProps, ...customPropsProvider(component, { ...nativeProps, ...stateBasedProps, ...callbackProps }), key },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
