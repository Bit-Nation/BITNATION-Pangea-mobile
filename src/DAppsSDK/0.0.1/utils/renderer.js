// @flow

import React from 'react';
import components from '../components';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => components[typeName];

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, custom: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const customProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (typeof propName !== 'string') {
      console.warn('Prop name is not a string');
      return;
    }

    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (validProps.custom.includes(propName)) {
      // It's an allowed custom prop.
      customProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, customProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === components.Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = components[type];

  if (component == null) {
    console.warn(`Trying to render unknown component type ${type}.`);
    return null;
  }

  if (component.validNativeProps != null && Array.isArray(component.validNativeProps) === false) {
    console.warn(`validNativeProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.customProps != null && Array.isArray(component.customProps) === false) {
    console.warn(`customProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.callbackProps != null && Array.isArray(component.callbackProps) === false) {
    console.warn(`callbackProps field of component ${type} should be an array of strings.`);
    return null;
  }

  const { props } = json;

  const { nativeProps = props, customProps = {}, callbackProps = {} } =
    validateProps(json.props, {
      native: component.validNativeProps || [],
      custom: component.customProps || [],
      callbacks: component.callbackProps || [],
    }, type);

  return React.createElement(
    component,
    {
      nativeProps,
      ...customProps,
      ...customPropsProvider(component, { ...nativeProps, ...customProps, ...callbackProps }),
      key,
    },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
