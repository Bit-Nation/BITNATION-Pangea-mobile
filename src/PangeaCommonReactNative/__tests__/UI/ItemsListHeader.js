import React from 'react';
import renderer from 'react-test-renderer';

import NationListHeader from '../../UI/ItemsListHeader';

test('ItemsListHeader renders correctly', () => {
  const tree = renderer
    .create(<NationListHeader title='Test header title' />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
