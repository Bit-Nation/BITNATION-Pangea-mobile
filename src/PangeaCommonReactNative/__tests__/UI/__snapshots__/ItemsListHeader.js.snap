// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsListHeader renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e5e5e5",
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingVertical": 15,
      "shadowColor": "#58595B",
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.8,
      "shadowRadius": 3,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "gray",
        "flex": 1,
        "fontFamily": "Roboto",
        "fontSize": 20,
        "fontWeight": "bold",
        "letterSpacing": 0,
        "lineHeight": 22,
        "marginLeft": 16,
      }
    }
  >
    Test header title
  </Text>
</View>
`;
